### DISCUSSION TITLE
Add an `apply` helper to the plugin closure API

### DISCUSSION THREAD

**Discussion by AlexVipond** [tailwindlabs/tailwindcss#3246] [Ideas]
I'm a big fan of writing simple custom plugins, powered by `@apply`, to add commonly needed component classes like `.btn` in each of my apps. My one small gripe is that the `@apply` syntax is a little funky/verbose—you have to write it like a media query with an empty object value:

```js
'.btn': {
  '@apply flex items-center space-x-2 px-2 py-1 rounded-4 text-3': {},
},
```

Intead, I use a really minimal function to provide a smoother API:

```js
function apply (classes) {
  return {
    [`@apply ${classes}`]: {}
  }
}

module.exports = {
  plugins: [
    plugin(({ addComponents }) => addComponents({
      '.btn': apply('flex items-center space-x-2 px-2 py-1 rounded-4 text-3'),
      ...otherCustomClasses,
    })
  ]
}
```

I propose adding that little `apply` helper to the `plugin` closure API as a quality-of-life improvement:

```js
module.exports = {
  plugins: [
    plugin(({ addComponents, apply }) => addComponents({
      '.btn': apply('flex items-center space-x-2 px-2 py-1 rounded-4 text-3'),
      ...otherCustomClasses,
    })
  ]
}
```

Via the spread syntax, it's also compatible with nested rules:

```js
module.exports = {
  plugins: [
    plugin(({ addComponents, apply }) => addComponents({
      '.btn': {
        ...apply('flex items-center space-x-2 px-2 py-1 rounded-4 text-3'),
        '.icon': { height: '1em', width: '1em' },
      }
      ...otherCustomClasses,
    })
  ]
}
```

I'm happy to do a pull request and add some tests if this seems like a good idea!

**Comment by AlexVipond** [comment]
For anyone who's still into this idea, I added an export to my `@baleada/tailwind-utilities` package so you can do it easily in your own custom Tailwind plugins.

Example below, and full docs are here: https://baleada.dev/docs/utilities#apply-helper

I also use this technique heavily in that same package. For ideas, check out the source: https://github.com/baleada/tailwind-utilities/blob/5cb2374c20cc60314e78cf64909cc0e9d0f89641/src/index.ts#L63-L66

```js
const { createApply } = require('@baleada/tailwind-utilities')
const plugin = require('tailwindcss/plugin')

const myPlugin = plugin(({ addUtilities, config }) => {
  // Use the `config` function from the plugin API to get the
  // user's prefix (if unconfigured, it's an empty string).
  const prefix = config('prefix')
  // Pass the prefix to the `createApply` function to create an `apply`
  // function for this specific Tailwind user's config.
  const apply = createApply(prefix)

  addUtilities({
    // Pass any string of Tailwind classes to `apply`
    '.center': apply(`
      absolute
      top-1/2 left-1/2
      -translate-x-1/2 -translate-y-1/2
    `)
  })
})
```

**Comment by tresorama** [comment]
Is this syntax in the docs ? I cannot find it
In case is not present, where do you find an explanation/example of the usage ?
```js
'.btn': {
  '@apply flex items-center space-x-2 px-2 py-1 rounded-4 text-3': {},
},
```

Thanks @AlexVipond for this insight.

**Reply by AlexVipond** [reply]
As far as I know, it's not in the docs, because it's more of a CSS-in-JS feature than a Tailwind feature. (Tailwind plugins are written with CSS-in-JS syntax, and this usage of `@apply` is just the way you do at-rules in that syntax.)

It also has some pitfalls like what I mentioned in my other comment—if the plugin is used in a project that configures a custom Tailwind class prefix, it will fail/error silently and won't register your utility or component classes.

But outside of those cases, it's a great trick and works the same way as this CSS:

```
@layer components {
  .btn {
    @apply ...
  }
}
```

This compiles to plain CSS with property/value pairs for each class you applied.

Use it anytime you want a nicer DX for writing plugins that add utilities or component classes, but just be aware of those pitfalls I mentioned.

**Reply by tresorama** [reply]
Thank you so much, I never tought of using "@apply ..." the same as "@media ...." in CSS-in-JS.
This means that Tailwind parse every at-rule from the CSS-in-JS object and "do some magic" when it encounter the proprietary "@apply", converting theme values.
But now make "some" sense.

Thanks also for the "prefix" insight.
I'm currently testing Tailwind config customization to decide which is the best route to use Tailwind with Gutenberg from Wordpress.
